core/src/main/scala/org/apache/spark/rdd/PipedRDD.scala

```diff
@@ -112,73 +101,199 @@ private[spark] class PipedRDD[T: ClassTag](
         workInTaskDirectory = true
       } catch {
         case e: Exception => logError("Unable to setup task working directory: " + e.getMessage +
-          " (" + taskDirectory + ")", e)
+            " (" + taskDirectory + ")", e)
       }
     }
 
     val proc = pb.start()
     val env = SparkEnv.get
+    val childThreadException = new AtomicReference[Throwable](null)
 
     // Start a thread to print the process's stderr to ours
-    new Thread("stderr reader for " + command) {
-      override def run() {
-        for (line <- Source.fromInputStream(proc.getErrorStream).getLines) {
-          // scalastyle:off println
-          System.err.println(line)
-          // scalastyle:on println
+    new Thread(s"stderr reader for $command") {
+      override def run(): Unit = {
+        val os = System.err
+        try {
+          ByteStreams.copy(proc.getErrorStream, os)
+        } catch {
+          case t: Throwable => childThreadException.set(t)
+        } finally {
+          os.close()
         }
       }
     }.start()
 
     // Start a thread to feed the process input from our parent's iterator
-    new Thread("stdin writer for " + command) {
-      override def run() {
+    new Thread(s"stdin writer for $command") {
+      override def run(): Unit = {
         TaskContext.setTaskContext(context)
-        val out = new PrintWriter(proc.getOutputStream)
+        val dos = new DataOutputStream(
+          new BufferedOutputStream(proc.getOutputStream, bufferSize))
+        try {
+          for (elem <- firstParent[I].iterator(split, context)) {
+            inputWriter.write(dos, elem)
+          }
+        } catch {
+          case t: Throwable => childThreadException.set(t)
+        } finally {
+          dos.close()
+        }
+      }
+    }.start()
 
-        // scalastyle:off println
-        // input the pipe context firstly
-        if (printPipeContext != null) {
-          printPipeContext(out.println(_))
+    val dis = new DataInputStream(
+      new BufferedInputStream(proc.getInputStream, bufferSize))
+    new Iterator[O] {
+      private var cached: O = _
+
+      def next(): O = {
+        if (!hasNext()) {
+          throw new NoSuchElementException()
         }
-        for (elem <- firstParent[T].iterator(split, context)) {
-          if (printRDDElement != null) {
-            printRDDElement(elem, out.println(_))
+
+        val elem = cached
+        cached = null.asInstanceOf[O]
+        elem
+      }
+
+      def hasNext(): Boolean = {
```

Comment:
The former implementation would be equivalent to dis.hasNext(). This is a change in behavior.